text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1412 @@
+
+As `Sized` is to `const Sized`, `ValueSized` is to `const ValueSized`: `ValueSized`
+requires a value, knowledge of the type and target platform, and can only be
+computed at runtime. For example, `[svint8_t]` requires a value to know how
+many elements there are, and then information from the runtime environment
+to know the size of a `svint8_t`.
+
+`Pointee` is implemented by any type that can be used behind a pointer, which is
+to say, every type (put otherwise, these types may or may not be sized at all).
+For example, `Pointee` is therefore implemented on a `u32` which is trivially
+sized, a `[usize]` which is dynamically sized, a `svint8_t` which is
+runtime-sized and an `extern type` (from [rfcs#1861][rfc_extern_types]) which has
+no known size.
+
+All type parameters have an implicit bound of `const Sized` which will be
+automatically removed if a `Sized`, `const ValueSized`, `ValueSized` or
+`Pointee` bound is present instead.
+
+Prior to the introduction of `ValueSized` and `Pointee`, `Sized`'s implicit bound
+(now a `const Sized` implicit bound) could be removed using the `?Sized` syntax,
+which is now equivalent to a `ValueSized` bound in non-`const fn`s and
+`~const ValueSized` in `const fn`s and will be deprecated in the next edition.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Introduce a new marker trait, `ValueSized`, adding it to a trait hierarchy with the
+[`Sized`][api_sized] and [`Pointee`][api_pointee] traits, and make `Sized` and `ValueSized`
+traits `const`:
+
+```
+    ┌────────────────┐                  ┌─────────────────────────────┐
+    │ const Sized    │ ───────────────→ │ Sized                       │
+    │ {type, target} │     implies      │ {type, target, runtime env} │
+    └────────────────┘                  └─────────────────────────────┘
+            │                                          │
+         implies                                    implies
+            │                                          │
+            ↓                                          ↓
+┌───────────────────────┐             ┌────────────────────────────────────┐
+│ const ValueSized      │ ──────────→ │ ValueSized                         │
+│ {type, target, value} │   implies   │ {type, target, runtime env, value} │
+└───────────────────────┘             └────────────────────────────────────┘
+                                                       │
+                                                    implies
+                                                       │
+                               ┌───────────────────────┘
+                               ↓
+                     ┌──────────────────┐
+                     │ Pointee          │
+                     │ {runtime env, *} │
+                     └──────────────────┘
+```
+
+Or, in Rust syntax:
+
+```rust=
+const trait Sized: ~const ValueSized {}
+
+const trait ValueSized: std::ptr::Pointee {}
```

Comment:
Note that this is a breaking change: It makes it ambiguous whether `T::Metadata` refers to the assoc type of `Pointee` or an user-provided trait:

```rust
trait Foo {
    type Metadata;
}

fn foo<T>() -> T::Metadata // does this refer to Foo::Metadata or Pointee::Metadata?
where
    T: Foo /* + Sized + ValueSized + Pointee*/
{
    todo!()
}
```

This can break [real code](https://github.com/search?q=lang%3Arust+%2F%28%3F-i%29%5BA-Z%5D%3A%3AMetadata%2F+&type=code) and should be addressed by this RFC.

Reply:
This is also why `Sized` (as it exists today) cannot have a `const SIZE: usize` associated const. Perhaps there is need for a `#[bikeshed_hide_associated_item]` attr applied to trait assocaited items so they can only be accessed using fully qualified syntax `<T as Pointee>::Metadata`; or perhaps the attr could be rustc-internal and indicate an edition before which the item is hidden as such.

Reply:
This is a good catch, I hadn't considered this. A simple alternative that earlier versions of the proposal had was just to introduce a new `Pointee` marker trait rather than re-use `std::ptr::Pointee`.

Reply:
There was anyway a proposal somewhere to expose the metadata type as `Metadata<Pointee>` rather than `<T as Pointee>::Metadata`, this would be another argument in favor of that.

Reply:
I've updated the RFC to make `Pointee` a new marker trait.